File: engine/src/dhan/market_data.rs
// Verified against actual Dhan API response on 2026-03-24.
// Response: {"status":"success","data":{"NSE_EQ":{"1333":{"last_price":755,"ohlc":{...},...}}}}
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use anyhow::Result;
use crate::dhan::client::DhanClient;

#[derive(Debug, Deserialize, Default, Clone)]
pub struct QuoteOhlc {
    #[serde(default)] pub open: f32,
    #[serde(default)] pub high: f32,
    #[serde(default)] pub low: f32,
    #[serde(default)] pub close: f32,  // previous day's EOD closing price
}

#[allow(dead_code)]
#[derive(Debug, Deserialize, Default, Clone)]
pub struct DepthLevel {
    #[serde(default)] pub price: f32,
    #[serde(default)] pub quantity: u32,
}

#[allow(dead_code)]
#[derive(Debug, Deserialize, Default, Clone)]
pub struct Depth {
    #[serde(default)] pub buy: Vec<DepthLevel>,
    #[serde(default)] pub sell: Vec<DepthLevel>,
}

#[derive(Debug, Deserialize, Default, Clone)]
pub struct QuoteItem {
    #[serde(default)] pub last_price: f32,
    #[serde(default)] pub ohlc: QuoteOhlc,
    #[serde(default)] pub volume: u64,
    #[allow(dead_code)] #[serde(default)] pub oi: u64,
    #[allow(dead_code)] #[serde(default)] pub buy_quantity: u32,
    #[allow(dead_code)] #[serde(default)] pub sell_quantity: u32,
    #[allow(dead_code)] #[serde(default)] pub depth: Depth,
}

impl QuoteItem {
    pub fn open(&self) -> f32 { self.ohlc.open }
    pub fn high(&self) -> f32 { self.ohlc.high }
    pub fn low(&self) -> f32 { self.ohlc.low }
    pub fn close(&self) -> f32 { self.ohlc.close }
}

#[derive(Debug, Serialize)]
struct QuoteRequest {
    #[serde(rename = "NSE_EQ", skip_serializing_if = "Vec::is_empty")]
    nse_eq: Vec<u64>,
}

/// Fetch quotes for a batch of security IDs (max 1000 per call).
/// Dhan requires security IDs sent as integers, not strings.
/// Retries up to 2 times on 429/5xx with exponential backoff.
pub async fn fetch_quotes(
    client: &DhanClient,
    security_ids: &[String],
    endpoint: &str,
) -> Result<HashMap<String, QuoteItem>> {
    let ids: Vec<u64> = security_ids.iter()
        .filter_map(|s| s.parse::<u64>().ok())
        .collect();
    let req_body = QuoteRequest { nse_eq: ids };

    let mut last_err = String::new();
    for attempt in 0..3u32 {
        if attempt > 0 {
            let backoff = std::time::Duration::from_secs(2u64.pow(attempt));
            tracing::warn!("[QUOTE] Retry {}/2 after {}s backoff", attempt, backoff.as_secs());
            tokio::time::sleep(backoff).await;
        }

        let resp = match client.post(endpoint).json(&req_body).send().await {
            Ok(r) => r,
            Err(e) => { last_err = format!("request error: {}", e); continue; }
        };

        let http_status = resp.status().as_u16();
        let text = resp.text().await.unwrap_or_default();

        if http_status == 429 {
            tracing::warn!("[QUOTE] 429 rate limited (attempt {})", attempt + 1);
            last_err = "429 rate limited".into();
            continue;
        }
        if http_status >= 500 {
            tracing::warn!("[QUOTE] Server error {} (attempt {})", http_status, attempt + 1);
            last_err = format!("server error {}", http_status);
            continue;
        }

        if client.debug {
            let preview: String = text.chars().take(500).collect();
            tracing::debug!("Quote API raw response: {}", preview);
        }

        let v: serde_json::Value = serde_json::from_str(&text).map_err(|e| {
            let preview: String = text.chars().take(300).collect();
            anyhow::anyhow!("JSON parse error: {} | body: {}", e, preview)
        })?;

        let status = v["status"].as_str().unwrap_or("unknown");
        if status != "success" {
            let msg: String = v["data"].to_string().chars().take(200).collect();
            return Err(anyhow::anyhow!("Dhan API status={} | {}", status, msg));
        }

        let mut result = HashMap::new();
        if let Some(segments) = v["data"].as_object() {
            for (_segment, items) in segments {
                if let Some(items_map) = items.as_object() {
                    for (sec_id, item_val) in items_map {
                        match serde_json::from_value::<QuoteItem>(item_val.clone()) {
                            Ok(q) => { result.insert(sec_id.clone(), q); }
                            Err(e) => {
                                tracing::warn!("Skipping sec_id={}: parse error: {}", sec_id, e);
                            }
                        }
                    }
                }
            }
        }
        return Ok(result);
    }

    Err(anyhow::anyhow!("fetch_quotes failed after 3 attempts: {}", last_err))
}

/// 1-minute intraday candle from Dhan intraday charts API.
#[derive(Debug, Deserialize)]
pub struct IntradayCandle {
    pub timestamp: i64,
    pub open: f32,
    pub high: f32,
    pub low: f32,
    pub close: f32,
    #[serde(default)]
    pub volume: f64,
}

/// Intraday historical response wrapper.
#[derive(Debug, Deserialize)]
pub struct IntradayResponse {
    #[serde(rename = "open", default)]
    pub open: Vec<f64>,
    #[serde(rename = "high", default)]
    pub high: Vec<f64>,
    #[serde(rename = "low", default)]
    pub low: Vec<f64>,
    #[serde(rename = "close", default)]
    pub close: Vec<f64>,
    #[serde(rename = "timestamp", default)]
    pub timestamp: Vec<i64>,
    #[serde(rename = "volume", default)]
    pub volume: Vec<f64>,
}

/// Fetch 1-minute intraday candles for a security between two dates.
/// Returns all candles; caller splits by date and converts to snapshot buckets.
pub async fn fetch_intraday_candles(
    client: &DhanClient,
    security_id: &str,
    from_date: &str, // "YYYY-MM-DD"
    to_date: &str,   // "YYYY-MM-DD"
) -> Result<IntradayResponse> {
    #[derive(Serialize)]
    struct Req<'a> {
        #[serde(rename = "securityId")]
        security_id: &'a str,
        #[serde(rename = "exchangeSegment")]
        exchange_segment: &'a str,
        instrument: &'a str,
        expiry_code: i32,
        #[serde(rename = "fromDate")]
        from_date: &'a str,
        #[serde(rename = "toDate")]
        to_date: &'a str,
    }

    let req_body = Req {
        security_id,
        exchange_segment: "NSE_EQ",
        instrument: "EQUITY",
        expiry_code: 0,
        from_date,
        to_date,
    };

    let mut last_err = String::new();
    for attempt in 0..3u32 {
        if attempt > 0 {
            let backoff = std::time::Duration::from_secs(2u64.pow(attempt));
            tokio::time::sleep(backoff).await;
        }

        let resp = match client.post("v2/charts/intraday").json(&req_body).send().await {
            Ok(r) => r,
            Err(e) => { last_err = format!("request error: {}", e); continue; }
        };

        let http_status = resp.status().as_u16();

        if http_status == 429 || http_status == 400 {
            let body = resp.text().await.unwrap_or_default();
            tracing::warn!("[INTRADAY] HTTP {} (attempt {}): {}", http_status, attempt + 1, &body[..body.len().min(200)]);
            last_err = format!("HTTP {}", http_status);
            continue;
        }
        if http_status >= 500 {
            last_err = format!("server error {}", http_status);
            continue;
        }

        let data: IntradayResponse = match resp.json().await {
            Ok(d) => d,
            Err(e) => { last_err = format!("json error: {}", e); continue; }
        };

        return Ok(data);
    }

    Err(anyhow::anyhow!("fetch_intraday_candles failed after 3 attempts: {}", last_err))
}

File: python-engine/src/swing_atlas/repositories/dhan/market_data.py
"""Dhan quote REST fetch -- mirrors engine/src/dhan/market_data.rs::fetch_quotes.

Retries up to 2 times (3 attempts total) on 429/5xx with exponential backoff,
matching the Rust original -- this feeds live paper-trade price enrichment and
the swing dashboard, both of which degrade to cached/historical prices on
failure rather than surfacing an error to the browser.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass, field
from typing import Any

import httpx

logger = logging.getLogger(__name__)

_MAX_ATTEMPTS = 3
_REQUEST_TIMEOUT_SECS = 15.0


@dataclass(frozen=True, slots=True)
class QuoteOhlc:
    open: float = 0.0
    high: float = 0.0
    low: float = 0.0
    close: float = 0.0  # previous day's EOD closing price


@dataclass(frozen=True, slots=True)
class QuoteItem:
    last_price: float = 0.0
    ohlc: QuoteOhlc = field(default_factory=QuoteOhlc)
    volume: int = 0


async def fetch_quotes(
    client: httpx.AsyncClient,
    base_url: str,
    access_token: str,
    client_id: str,
    security_ids: list[str],
    endpoint: str,
) -> dict[str, QuoteItem]:
    """Fetch quotes for a batch of security IDs (max 1000 per call). Dhan
    requires security IDs sent as integers, not strings."""
    ids = [int(sid) for sid in security_ids if sid.isdigit()]
    url = f"{base_url.rstrip('/')}{endpoint}"
    headers = {
        "access-token": access_token,
        "client-id": client_id,
        "Content-Type": "application/json",
    }

    last_err = ""
    for attempt in range(_MAX_ATTEMPTS):
        if attempt > 0:
            backoff = 2**attempt
            logger.warning("[QUOTE] Retry %d/2 after %ds backoff", attempt, backoff)
            await asyncio.sleep(backoff)

        try:
            response = await client.post(
                url, json={"NSE_EQ": ids}, headers=headers, timeout=_REQUEST_TIMEOUT_SECS
            )
        except httpx.HTTPError as exc:
            last_err = f"request error: {exc}"
            continue

        if response.status_code == 429:
            logger.warning("[QUOTE] 429 rate limited (attempt %d)", attempt + 1)
            last_err = "429 rate limited"
            continue
        if response.status_code >= 500:
            logger.warning(
                "[QUOTE] Server error %d (attempt %d)", response.status_code, attempt + 1
            )
            last_err = f"server error {response.status_code}"
            continue

        try:
            payload = response.json()
        except ValueError as exc:
            preview = response.text[:300]
            raise RuntimeError(f"JSON parse error: {exc} | body: {preview}") from exc

        status = payload.get("status", "unknown")
        if status != "success":
            message = str(payload.get("data", ""))[:200]
            raise RuntimeError(f"Dhan API status={status} | {message}")

        return _parse_quote_segments(payload.get("data") or {})

    raise RuntimeError(f"fetch_quotes failed after {_MAX_ATTEMPTS} attempts: {last_err}")


def _parse_quote_segments(segments: dict[str, Any]) -> dict[str, QuoteItem]:
    result: dict[str, QuoteItem] = {}
    for items in segments.values():
        if not isinstance(items, dict):
            continue
        for security_id, item in items.items():
            try:
                ohlc = item.get("ohlc") or {}
                result[security_id] = QuoteItem(
                    last_price=float(item.get("last_price", 0.0)),
                    ohlc=QuoteOhlc(
                        open=float(ohlc.get("open", 0.0)),
                        high=float(ohlc.get("high", 0.0)),
                        low=float(ohlc.get("low", 0.0)),
                        close=float(ohlc.get("close", 0.0)),
                    ),
                    volume=int(item.get("volume", 0)),
                )
            except (TypeError, ValueError) as exc:
                logger.warning("Skipping sec_id=%s: parse error: %s", security_id, exc)
    return result


@dataclass(frozen=True, slots=True)
class IntradayResponse:
    """1-minute intraday candles from Dhan's charts API, in Dhan's own
    parallel-array wire format (not an array of candle objects)."""

    open: list[float] = field(default_factory=list)
    high: list[float] = field(default_factory=list)
    low: list[float] = field(default_factory=list)
    close: list[float] = field(default_factory=list)
    timestamp: list[float] = field(default_factory=list)
    volume: list[float] = field(default_factory=list)


async def fetch_intraday_candles(
    client: httpx.AsyncClient,
    base_url: str,
    access_token: str,
    client_id: str,
    security_id: str,
    from_date: str,
    to_date: str,
) -> IntradayResponse:
    """Fetch 1-minute intraday candles for a security between two dates
    ("YYYY-MM-DD HH:MM:SS"). Retries up to 2 times (3 attempts total) on
    429/400/5xx with exponential backoff, matching the Rust original."""
    url = f"{base_url.rstrip('/')}/charts/intraday"
    headers = {
        "access-token": access_token,
        "client-id": client_id,
        "Content-Type": "application/json",
    }
    body = {
        "securityId": security_id,
        "exchangeSegment": "NSE_EQ",
        "instrument": "EQUITY",
        "interval": "1",
        "oi": False,
        "fromDate": from_date,
        "toDate": to_date,
    }

    last_err = ""
    for attempt in range(_MAX_ATTEMPTS):
        if attempt > 0:
            await asyncio.sleep(2**attempt)

        try:
            response = await client.post(
                url, json=body, headers=headers, timeout=_REQUEST_TIMEOUT_SECS
            )
        except httpx.HTTPError as exc:
            last_err = f"request error: {exc}"
            continue

        preview = response.text[:300]
        if response.status_code in (429, 400):
            logger.warning(
                "[INTRADAY] HTTP %d (attempt %d): %s", response.status_code, attempt + 1, preview
            )
            last_err = f"HTTP {response.status_code}: {preview}"
            continue
        if response.status_code >= 500:
            last_err = f"server error {response.status_code}: {preview}"
            continue

        try:
            payload = response.json()
        except ValueError as exc:
            last_err = f"json error: {exc} | body: {preview}"
            continue

        return IntradayResponse(
            open=[float(v) for v in payload.get("open", [])],
            high=[float(v) for v in payload.get("high", [])],
            low=[float(v) for v in payload.get("low", [])],
            close=[float(v) for v in payload.get("close", [])],
            timestamp=[float(v) for v in payload.get("timestamp", [])],
            volume=[float(v) for v in payload.get("volume", [])],
        )

    raise RuntimeError(f"fetch_intraday_candles failed after {_MAX_ATTEMPTS} attempts: {last_err}")
